ExtrudeGeometry: Fix regression introduced by #30750.

mlasala45 opened the pull request:
Fixed #30819

**Description**

Changes introduced by #30750 resulted in the inputs to ShapeUtils.triangulateGeometry not being initialized when bevelSegments == 0 (which happens when bevelEnabled == false), causing triangulation to fail. This PR adds code to manually populate those lists if they are empty before triangulation, fixing the issue. See related issue for examples.

mlasala45 committed:
Fix vertex lists not being generated correctly when bevelSegments == 0

mlasala45 commented:
Maybe it would be worth having a test screenshot for bevelEnabled == false?

Mugen87 commented on src/geometries/ExtrudeGeometry.js:
```diff
@@ -439,6 +439,27 @@ class ExtrudeGeometry extends BufferGeometry {
 			}
 
+			//When bevelSegments == 0, the modified vert arrays will not have been populated. We do this here.
+			{
```
Can we avoid using these extra brackets?

mlasala45 committed:
Removed superfluous scope block around new code

WestLangley commented:
These new changes appear to be causing additional regressions. Also check the console errors.

<img width="633" alt="Screenshot 2025-03-29 at 7 25 29 PM" src="https://github.com/user-attachments/assets/29a273e4-a191-4786-a32a-7c5ac37292f8" />

https://threejs.org/examples/webgl_geometry_text.html // tested after merging this PR

I would be inclined to revert #30750 at this point and allow time for proper testing.

mlasala45 commented:
@WestLangley not seeing console errors, but I see the geometry bug (oddly, it's different on my end). I'll investigate.

mlasala45 committed:
Fixed using the wrong variable to populate contractedContourVertices

The variable "vertices" could probably use renaming for clarity.

mlasala45 commented:
Fixed. The change was using the wrong variable to populate contractedContourVertices, which would cause an error for shapes with holes when bevelEnabled == false. My bad, I should have tested the PR more thoroughly.

WestLangley commented:
>@WestLangley not seeing console errors, ...

The console errors occurred when the font was changed.

Unrelated to this PR, but FYI, file `8.svg` from the `webgl_loader_svg.html` example does not extrude properly. It is just a rectangle with a single hole.

mlasala45 commented:
> Unrelated to this PR, but FYI, file 8.svg from the webgl_loader_svg.html example does not extrude properly. It is just a rectangle with a single hole.
It appears that the outside boundary is being classified as a hole, while the inside boundary is being classified as a contour. This is happening before it's passed to ExtrudeGeometry, so it's probably a defect with the SVG file itself, the SVGLoader, or path.toShapes(). The issue seems to predate #30750 though.

WestLangley commented:
@gkjohnson I have done the best I can to identify the failure points. Someone other than myself will have to approve the code changes. 🙏

gkjohnson commented on src/geometries/ExtrudeGeometry.js:
```diff
@@ -439,6 +439,24 @@ class ExtrudeGeometry extends BufferGeometry {
 
 			}
 
+			//When bevelSegments == 0, the modified vert arrays will not have been populated. We do this here.
```
Lets use spaces after comments so we're consistent with the rest of the files.

gkjohnson commented on src/geometries/ExtrudeGeometry.js:
```diff
@@ -439,6 +439,24 @@ class ExtrudeGeometry extends BufferGeometry {
+
+			}
+
+			if ( expandedHoleVertices.length == 0 ) {
+
+				for ( let h = 0, hl = numHoles; h < hl; h ++ ) {
+
+					expandedHoleVertices.push( holes[ h ] ); //triangulateShape expects a Vec2[][], ie. an array of holes, with each hole being a Vec2[]
+
+				}
+
+			}
```
Is there ever a case where contract contour will have length 0 but expandedHoleVertices will not? I think for clarity we should handle these in the same condition.

gkjohnson commented on src/geometries/ExtrudeGeometry.js:
```diff
@@ -439,6 +439,24 @@ class ExtrudeGeometry extends BufferGeometry {
+
+			if ( contractedContourVertices.length == 0 ) {
+
+				contractedContourVertices.push( ...contour );
```
I've seen the spread operator cause stackoverflow errors when used with "push" - lets avoid using it here. Is there a reason to push the vertices onto the array rather than replacing the variable?

gkjohnson commented:
> I have done the best I can to identify the failure points. Someone other than myself will have to approve the code changes. 🙏

Sorry I missed that there were already other reviews on this PR somehow. I'm not an expert in this file but as this PR is addressing it looks like the tringulated arrays were not being populated if there were no bevel points. Using the "contour" and "holes" arrays should revert the behavior back to what it was before #30750 when bevels were disabled. This change also seems to work.

```js
let faces;
if ( contractedContourVertices.length === 0 ) {

	faces = ShapeUtils.triangulateShape( contour, holes );

} else {

	faces = ShapeUtils.triangulateShape( contractedContourVertices, expandedHoleVertices );

}
```

mlasala45 committed:
Restructured fix for efficiency and clarity

Now branches based on bevelSegments ?= 0, bypassing contractedContourVertices/expandedHoleVertices  generation when there is no bevel.

mlasala45 commented:
>Using the "contour" and "holes" arrays should revert the behavior back to what it was before https://github.com/mrdoob/three.js/pull/30750 when bevels were disabled.

@gkjohnson Good call. I've restructured this fix to check bevelSegments early, and avoid calculating contractedContourVertices and expandedHoleVertices  when not necessary. The diffs show up poorly due to indentation changes, but the structure is now:
```js
let faces;
if ( bevelSegments == 0 ) {

	faces = ShapeUtils.triangulateShape( contour, holes );

} else {
	// Calculate contractedContourVertices and expandedHoleVertices
	// ...

	faces = ShapeUtils.triangulateShape( contractedContourVertices, expandedHoleVertices );
}
```

gkjohnson reviewed (commented):
I think this looks good and ready to merge.

Mugen87 committed:
Update ExtrudeGeometry.js

Mugen87 committed:
Update ExtrudeGeometry.js
